File: composeApp/src/androidMain/kotlin/org/example/project/presentations/utils/TimeFormat.kt
package org.example.project.presentations.utils

import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.LocalDate
import kotlinx.datetime.LocalTime
import kotlinx.datetime.TimeZone
import kotlinx.datetime.isoDayNumber
import kotlinx.datetime.minus
import kotlinx.datetime.plus
import kotlinx.datetime.toLocalDateTime
import org.example.project.data.remote.dto.week_schedule.CourseClass
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlin.time.Clock

val today: LocalDate
    get() = Clock.System.now()
        .toLocalDateTime(TimeZone.currentSystemDefault()).date

fun Long.toFormatTime(): String {
    val formatter = DateTimeFormatter.ofPattern("hh:mm a", Locale.getDefault())

    return Instant.ofEpochMilli(this)
        .atZone(ZoneId.systemDefault())
        .format(formatter)
}

fun getTodayDayOfWeek(): Int = today.dayOfWeek.isoDayNumber

private fun getWeekOf(date: LocalDate): Pair<String, String> {
    val dayOfWeek = date.dayOfWeek.isoDayNumber
    val start = date.minus(dayOfWeek - 1, DateTimeUnit.DAY)
    val end = date.plus(7 - dayOfWeek, DateTimeUnit.DAY)
    return Pair(start.toString(), end.toString())
}

fun getCurrentWeek() = getWeekOf(today)
fun getNextWeek(date: LocalDate) = getWeekOf(date.plus(7, DateTimeUnit.DAY))
fun getPreviousWeek(date: LocalDate) = getWeekOf(date.minus(7, DateTimeUnit.DAY))

fun String.toHourMinute(): String {
    val time = LocalTime.parse(this)
    return "%02d:%02d".format(time.hour, time.minute)
}

fun String.toHourMinuteAmPm(): String {
    val time = LocalTime.parse(this)

    val hour12 = when {
        time.hour == 0 -> 12
        time.hour > 12 -> time.hour - 12
        else -> time.hour
    }

    val amPm = if (time.hour < 12) "AM" else "PM"

    return "%02d:%02d %s".format(hour12, time.minute, amPm)
}

fun CourseClass.isGoing(
    currentTime: LocalTime = Clock.System.now()
        .toLocalDateTime(TimeZone.currentSystemDefault())
        .time
): Boolean {
    val start = LocalTime.parse(startTime)
    val end = LocalTime.parse(endTime)

    return currentTime in start..end
}

fun CourseClass.getStatusText(
    currentTime: LocalTime = Clock.System.now()
        .toLocalDateTime(TimeZone.currentSystemDefault())
        .time
): String {
    val start = LocalTime.parse(startTime)
    val end = LocalTime.parse(endTime)

    val startMinutes = start.hour * 60 + start.minute
    val endMinutes = end.hour * 60 + end.minute
    val currentMinutes = currentTime.hour * 60 + currentTime.minute

    return when {
        currentMinutes > endMinutes -> {
            "Đã kết thúc"
        }

        currentMinutes in startMinutes..endMinutes -> {
            "Đang diễn ra"
        }

        startMinutes - currentMinutes <= 60 -> {
            val diff = startMinutes - currentMinutes
            "Sau $diff phút"
        }

        else -> {
            "Sắp diễn ra"
        }
    }
}

fun String.toSlashDate(): String {
    return this.replace("-", "/")
}

//date
fun String.toDisplayDate(): String {
    val date = LocalDate.parse(this)
    return "%02d/%02d/%04d".format(date.dayOfMonth, date.monthNumber, date.year)
} // "21-01-2022"

File: iosApp/iosApp/Core/Extensions/TimeFormat.swift
//
//  TimeFormat.swift
//  iosApp
//
//  Created by Tuan Nguyen on 3/30/26.
//

import Foundation
import Shared

var today: Date {
    return Date()
}

func getTodayDayOfWeek() -> Int {
    let calendar = Calendar.current
    let weekday = calendar.component(.weekday, from: today)
    
    // Swift: 1 = Sunday → convert sang ISO (1 = Monday)
    return weekday == 1 ? 7 : weekday - 1
}

extension Int64 {
    func toFormatTime() -> String {
        let date = Date(timeIntervalSince1970: TimeInterval(self) / 1000)
        
        let formatter = DateFormatter()
        formatter.dateFormat = "hh:mm a"
        
        return formatter.string(from: date)
    }
}


func getWeekOf(date: Date) -> (String, String) {
    let calendar = Calendar.current
    
    let weekday = getTodayDayOfWeek()
    
    let start = calendar.date(byAdding: .day, value: -(weekday - 1), to: date)!
    let end = calendar.date(byAdding: .day, value: (7 - weekday), to: date)!
    
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyy-MM-dd"
    
    return (formatter.string(from: start), formatter.string(from: end))
}

func getCurrentWeek() -> (String, String) {
    return getWeekOf(date: today)
}

func getNextWeek(date: Date) -> (String, String) {
    let next = Calendar.current.date(byAdding: .day, value: 7, to: date)!
    return getWeekOf(date: next)
}

func getPreviousWeek(date: Date) -> (String, String) {
    let prev = Calendar.current.date(byAdding: .day, value: -7, to: date)!
    return getWeekOf(date: prev)
}

extension CourseClass {
    func isGoing(currentTime: Date = Date()) -> Bool {
        let calendar = Calendar.current
        
        let now = calendar.component(.hour, from: currentTime) * 60 +
                  calendar.component(.minute, from: currentTime)
        
        let startParts = startTime.split(separator: ":")
        let endParts = endTime.split(separator: ":")
        
        guard let sh = Int(startParts[0]),
              let sm = Int(startParts[1]),
              let eh = Int(endParts[0]),
              let em = Int(endParts[1]) else {
            return false
        }
        
        let start = sh * 60 + sm
        let end = eh * 60 + em
        
        return now >= start && now <= end
    }
    
    func getStatusText(currentTime: Date = Date()) -> String {
            let calendar = Calendar.current
            
            let currentMinutes =
                calendar.component(.hour, from: currentTime) * 60 +
                calendar.component(.minute, from: currentTime)
            
            let startParts = startTime.split(separator: ":")
            let endParts = endTime.split(separator: ":")
            
            guard let sh = Int(startParts[0]),
                  let sm = Int(startParts[1]),
                  let eh = Int(endParts[0]),
                  let em = Int(endParts[1]) else {
                return ""
            }
            
            let startMinutes = sh * 60 + sm
            let endMinutes = eh * 60 + em
            
            if currentMinutes > endMinutes {
                return "Đã kết thúc"
            } else if currentMinutes >= startMinutes && currentMinutes <= endMinutes {
                return "Đang diễn ra"
            } else if startMinutes - currentMinutes <= 60 {
                let diff = startMinutes - currentMinutes
                return "Sau \(diff) phút"
            } else {
                return "Sắp diễn ra"
            }
        }
}
